bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java

```diff
@@ -0,0 +1,217 @@
+
+        private ComponentModel(final RecordComponent component, final CodecRegistry codecRegistry, final int index) {
+            this.component = component;
+            this.codec = computeCodec(component, codecRegistry);
+            this.index = index;
+            this.fieldName = computeFieldName(component);
+        }
+
+        String getComponentName() {
+            return component.getName();
+        }
+
+        String getFieldName() {
+            return fieldName;
+        }
+
+        Object getValue(final Record record) throws InvocationTargetException, IllegalAccessException {
+            return component.getAccessor().invoke(record);
+        }
+
+        private static Codec<?> computeCodec(final RecordComponent component, final CodecRegistry codecRegistry) {
+            var codec = codecRegistry.get(toWrapper(component.getType()));
+            var bsonRepresentationAnnotation = component.getAnnotation(BsonRepresentation.class);
+            if (bsonRepresentationAnnotation != null) {
+                if (codec instanceof RepresentationConfigurable<?> representationConfigurable) {
+                    codec = representationConfigurable.withRepresentation(bsonRepresentationAnnotation.value());
+                } else {
+                    throw new CodecConfigurationException(
+                            format("Codec for %s must implement RepresentationConfigurable to support BsonRepresentation",
+                                    codec.getEncoderClass()));
+                }
+            }
+            return codec;
+        }
+
+        private static String computeFieldName(final RecordComponent component) {
+            if (component.isAnnotationPresent(BsonId.class)) {
+                return "_id";
+            } else if (component.isAnnotationPresent(BsonProperty.class)) {
+                return component.getAnnotation(BsonProperty.class).value();
+            }
+            return component.getName();
+        }
+    }
+
+    RecordCodec(final Class<T> clazz, final CodecRegistry codecRegistry) {
+        this.clazz = notNull("class", clazz);
+        canonicalConstructor = notNull("canonicalConstructor", getCanonicalConstructor(clazz));
+        componentModels = getComponentModels(clazz, codecRegistry);
+        fieldNameToComponentModel = componentModels.stream()
+                .collect(Collectors.toMap(ComponentModel::getFieldName, Function.identity()));
+        componentModelForId = getComponentModelForId(clazz, componentModels);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T decode(final BsonReader reader, final DecoderContext decoderContext) {
+        reader.readStartDocument();
+
+        Object[] constructorArguments = new Object[componentModels.size()];
+        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+            var fieldName = reader.readName();
+            var componentModel = fieldNameToComponentModel.get(fieldName);
```

Comment:
[performance-related mumbling, no action is expected]

As far as I remember (and I may be completely wrong here), some of the optimizations that @BorisDog prototyped for POCOs in the C# driver are:
- Rely on the order in which fields are serialized when decoding: instead of looking up in a map, one may simply pick the next component model from `componentModels` (this implies that `componentModelForId` is excluded form `componentModels`).
- Since `null` values are not serialized, the previous optimization may result in the need to skip some of the `componentModels`. This means that the `ComponentModel.getFieldName` still needs to be checked agains the `fieldName` from the `reader`, however, such a comparison is less resource-consuming than `fieldNameToComponentModel.get`, because the latter does the same check in addition to other activities. Such a comparison may be further speed up by storing a BSON binary representation of the field name in `ComponentModel`, and then comparing it with a binary representation of the field name from `BsonReader` instead of decoding that binary representation into a `String` and then comparing strings (our `BsonReader` does not seem to support doing so).